src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOQuerySupport.scala

```diff
@@ -0,0 +1,124 @@
+    client.prepareSearch(indexname)
+      .setQuery(ESConstantScore(ESFilter(ESBool(ESMust(qmseq)))).toJson.compactPrint)
+      .setFrom(from)
+      .setSize(size)
+  }
+
+  def buildMainQuery(client: TransportClient, indexname: String, baseQuery: Seq[QueryMap], criteria: LibrarySearchParams): SearchRequestBuilder = {
+    val mainQuery = createESSearchRequest(client, indexname, baseQuery, criteria.from, criteria.size)
+    if (criteria.fieldAggregations.nonEmpty && criteria.fieldAggregations.size != criteria.searchFields.size) {
+      // if we are not collecting aggregation data (in the case of pagination)
+      // and if there are some aggregations that are not part of the search criteria
+      // then the aggregation data for those fields will be accurate from the main search
+      addAggregationsToQuery(mainQuery, criteria.fieldAggregations.diff(criteria.searchFields.keySet.toIndexedSeq), criteria.maxAggregations)
+    }
+    mainQuery
+  }
+
+
+  def buildAggregateQueries(client: TransportClient, indexname: String, baseQuery: Seq[QueryMap], criteria: LibrarySearchParams): Seq[SearchRequestBuilder] = {
+    // for aggregtions fields that are part of the current search criteria, we need to do a separate
+    // aggregate request *without* that term in the search criteria
+    (criteria.searchFields.keys map { field =>
+      val query = baseQuery.filterNot((m: QueryMap) => m.isInstanceOf[ESBool] && m.asInstanceOf[ESBool].bool.isInstanceOf[ESShould]
+        && m.asInstanceOf[ESBool].bool.asInstanceOf[ESShould].should.last.isInstanceOf[ESTerm]
+        && m.asInstanceOf[ESBool].bool.asInstanceOf[ESShould].should.last.asInstanceOf[ESTerm].term.last._1.equals(field))
+      // setting size to 0, we will ignore the actual search results
+      addAggregationsToQuery(createESSearchRequest(client, indexname, query, 0, 0), Seq(field), criteria.maxAggregations)
+    }).toSeq
+  }
+
+  def getAggregationsFromResults(aggResults: Aggregations): Seq[LibraryAggregationResponse] = {
+    if (aggResults == null)
+      Seq.empty
+    else {
+      aggResults.getAsMap.keySet().asScala.toSeq map { field: String =>
+        val terms: Terms = aggResults.get(field)
+        LibraryAggregationResponse(terms.getName,
+          AggregationFieldResults(terms.getSumOfOtherDocCounts.toInt,
+            terms.getBuckets.asScala map { bucket: Terms.Bucket =>
+              AggregationTermResult(bucket.getKey.toString, bucket.getDocCount.toInt)
+            }))
+      }
+    }
+  }
+
+
+  def findDocumentsWithAggregateInfo(client: TransportClient, indexname: String, criteria: LibrarySearchParams): Future[LibrarySearchResponse] = {
+    val baseQuery = createQuery(criteria)
+    val mainQuery = buildMainQuery(client, indexname, baseQuery, criteria)
+    val aggregateQueries = buildAggregateQueries(client, indexname, baseQuery, criteria)
```

Comment:
nit: be consistent with naming here, especially since this is pretty complicated. `aggregateQueries` turns into `aggFutures`; good. `mainQuery` turns into `searchFuture`, which got confusing. Rename either the query or the future?